core/src/main/java/org/apache/calcite/rex/RexShuttle.java

```diff
@@ -160,8 +161,8 @@ public RexNode visitCall(final RexCall call) {
    */
   public void visitList(
       List<? extends RexNode> exprs, List<RexNode> outExprs) {
-    for (RexNode expr : exprs) {
-      outExprs.add(expr.accept(this));
+    for (int i = 0; i < exprs.size(); i++) {
+      outExprs.add(exprs.get(i).accept(this));
```

Comment:
If the `ImmutableList` iteration has bad performance, maybe we should promote and fix `ImmutableList`, then more code would benefit from that fix.

Reply:
There is nothing wrong with ImmutableList. It is just not a good fit here.